cranelift/codegen/src/incremental_cache.rs

```diff
@@ -0,0 +1,788 @@
+    ///
+    /// See comment on top of `CacheKey`.
+    fn check_from_src(self) -> Function {
+        Function {
+            version_marker: self.version_marker,
+            name: Default::default(), // caching is resilient to source locs
+            signature: self.signature,
+            sized_stack_slots: self.sized_stack_slots,
+            dynamic_stack_slots: self.dynamic_stack_slots,
+            global_values: self.global_values,
+            heaps: self.heaps,
+            tables: self.tables,
+            jump_tables: self.jump_tables,
+            dfg: self.dfg.check_from_src(),
+            layout: self.layout,
+            srclocs: Default::default(), // caching is resilient to source locs
+            stack_limit: self.stack_limit,
+        }
+    }
+
+    /// Creates a new cache store key for a function.
+    ///
+    /// This is a bit expensive to compute, so it should be cached and reused as much as possible.
+    fn new(isa: &dyn TargetIsa, f: &Function, offset: SourceLoc) -> Self {
+        let constants = CachedConstantPool {
+            handles_to_values: f.dfg.constants.handles_to_values.clone(),
+            values_to_handles: f
+                .dfg
+                .constants
+                .values_to_handles
+                .iter()
+                .map(|(k, v)| (k.clone(), v.clone()))
+                .collect(),
+        };
+
+        let values_labels = f.dfg.values_labels.clone().map(|vl| {
+            vl.into_iter()
+                .map(|(k, v)| {
+                    let v = match v {
+                        ValueLabelAssignments::Starts(vec) => CachedValueLabelAssignments::Starts(
+                            vec.into_iter()
+                                .map(|vls| CachedValueLabelStart {
+                                    from: RelSourceLoc::new(vls.from, offset),
+                                    label: vls.label,
+                                })
+                                .collect(),
+                        ),
+                        ValueLabelAssignments::Alias { from, value } => {
+                            CachedValueLabelAssignments::Alias {
+                                from: RelSourceLoc::new(from, offset),
+                                value,
+                            }
+                        }
+                    };
+                    (k, v)
+                })
+                .collect()
+        });
+
+        let ext_funcs = f
+            .dfg
+            .ext_funcs
+            .values()
+            .map(|ext_data| {
+                let name = match ext_data.name {
+                    ExternalName::User { .. } => {
+                        // Remove the identifying properties of the call to that external function,
+                        // as we want two functions with calls to different functions to reuse each
+                        // other's cached artifacts.
+                        CachedExternalName::User
```

Comment:
Won't this cause miscompilations when you at first call functions a, a, b and the next time a, b, b as there is no way to distinguish them based on the cache key?

Reply:
That's the thing: there's a perfect bijection between Func's `FuncRef` and `ExternalName`, so we don't need to record the external names, and we can just be "resilient" with them, meaning that it's fine to the actual callees change. This will end up generating different `MachReloc` later.

I think Chris' suggestion to split the `Function`/`MachCompileResult` into the "core" and parameters ought to make this clearer and easier to understand.

Reply:
> That's the thing: there's a perfect bijection between Func's FuncRef and ExternalName, so we don't need to record the external names, and we can just be "resilient" with them, meaning that it's fine to the actual callees change. This will end up generating different MachReloc later.

I don't see how a bijection helps.

Reply:
(... so we might as well just use the `FuncRef` everywhere directly, to avoid maintaining the invariant that we want this bijection; and this is in line with what Chris suggested in his comment, so I'm doing that now.)

Reply:
> When we load from the cache, and have a function with different FuncRefs -> ExternalNames, we can:

But what if you have a function with the same FuncRef -> ExternalName, but where the body eg calls functions in a different order?

Reply:
Each different function call (`ir::Opcode::Call`) references the `FuncRef` it's calling, so this is entirely determined by the CLIF IR. If two functions have the same body and each call two functions, but two different functions, then two different `FuncRef` are involved, leading to different lookups in the `FuncRef -> ExternalName` mapping, thus different `ExternalName`s. (Well, that's assuming that the `ExternalName`s are all different from each other, which would be important if we stayed down that path but that's not the case.)

Reply:
You are normalizing away the difference between different ExternalName's away here, right? `CachedExternalName::User` doesn't have any field to distinguish one from another.

Reply:
Indeed, because they're "spurious", at this point; only the `FuncRef` really matters, and there's a way to go from `ExternalName` back to `FuncRef` (in the _source_ function that filled the cache), and then from the `FuncRef` to `ExternalName` (in the function for which we're reusing the cache).

But don't get too attached to this, it's going to be simplified by Chris' suggestion below :)